tsdb/importer/importer.go

```diff
@@ -0,0 +1,407 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importer
+
+import (
+	"context"
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/pkg/value"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+	labels2 "github.com/prometheus/prometheus/tsdb/labels"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+)
+
+// Implementing the error interface to create a
+// constant, which cannot be overridden.
+// https://dave.cheney.net/2016/04/07/constant-errors
+type Error string
+
+func (e Error) Error() string {
+	return string(e)
+}
+
+const (
+	// This error is thrown when the current entry does not have a timestamp associated.
+	NoTimestampError = Error("expected timestamp with metric")
+	// This error is thrown when the sample being parsed currently aligns with the database's
+	// time ranges, but we cannot find a suitable block for it.
+	NoBlockFoundError = Error("no corresponding block found for current sample")
+	// This error is thrown when the sample being parsed currently has a corresponding block
+	// in the database, but has no metadata collected for it, curiously enough.
+	NoBlockMetaFoundError = Error("no metadata found for current samples' block")
+)
+
+type blockTimestampPair struct {
+	start, end int64
+}
+
+type newBlockMeta struct {
+	index      int
+	count      int
+	mint, maxt int64
+	isAligned  bool
+	dirs       []string
+}
+
+// Content Type for the Open Metrics Parser.
+// Needed to init the text parser.
+const contentType = "application/openmetrics-text; version=0.0.1; charset=utf-8"
+
+// ImportFromFile imports data from a file formatted according to the Open Metrics format,
+// converts it into block(s), and places the newly created block(s) in the
+// TSDB DB directory, where it is treated like any other block.
+func ImportFromFile(filePath string, dbPath string, maxSamplesInMemory int, logger log.Logger) error {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	dbMint, dbMaxt, blockTimes, err := getDbTimes(dbPath)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// TODO (dipack95): Read input file possibly by mmap'ing, and by chunking?
```

Comment:
I was thinking of maybe getting this PR accepted first, and then working on streaming the input data, and also re-using the allocated memory on each pass like you suggested in another comment. What do you think?

Reply:
I think implementing the streaming approach will require quite a few changes in the way that we conduct multiple passes over the data as well, which could take some time to get right, potentially delaying this PR even further.

Reply:
with the idea I had I think it will be just few extra lines so I think it is worth trying it in this PR.

Reply:
Okay, I'll give it a shot.

Reply:
So the issue I'm facing with streaming right now is trying to figure out how to segment the input data into blocks that each contain a single metrics bytes.

Using the code snippet you wrote, it divides the input file into blocks of 4096 bytes, which causes one of the entries to be incomplete, breaking the parser, which means that the scanner needs to fetch more of the file, and the process repeats, until the entire file is loaded anyway into the scanner. 

As far as I can tell, we will have to read the entire file anyway to figure out where we can safely create breaks for the parsers to do their thing.

Reply:
in the split function you can return `return 0, nil, nil` until parser returns a full record.
See: https://medium.com/golangspec/in-depth-introduction-to-bufio-scanner-in-golang-55483bb689b4

Reply:
Oh yeah, I did that - it doesn't fit our needs specifically, cause there's no way to roll back the scanner once it goes past a full record. For ex., if we have text like this:

```
# HELP metricA foo
# TYPE metricA gauge
metricA{foo="bar"} 1 100
metricA{foo="bar"} 2 101
metricA{foo="bar"} 3 103
<...>
# HELP metricB foo
# TYPE metricB gauge
metricB{foo="bar"} 1 101
metricB{foo="bar"} 2 102
# EOF
```

If we run the scanner over this, it will never cleanly be able to split them, without manual intervention. And once we use that `0, nil, nil` technique to fetch more data into the buffer, it might get us some of metric B's entries too, with there being no way to walk back the scanner's advances, so to speak.

I think might have to use a custom parser and splitter, to break the bytes into clean metric records for each metric.

Reply:
seems that the advance param is to tell the scanner how far to move in the reader. This is also described in that article.

Reply:
I guess it might some code changes to the scanner, but these are very well justified for this new feature so it is ok to change it if needed.

Reply:
Turns out I was just using the scanner incorrectly. :)
I've managed to get streaming working - please take a look @krasi-georgiev